### PR TITLE
feat: integrate relay

### PR DESCRIPTION
**Pull request by cartcrom**
to generate relay queries: yarn relay (or just yarn install)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -57,6 +63,7 @@
   "devDependencies": {
     "@ethersproject/experimental": "^5.4.0",
     "@graphql-codegen/cli": "1.21.5",
+    "@graphql-codegen/schema-ast": "^2.5.1",
     "@graphql-codegen/typescript": "1.22.3",
     "@graphql-codegen/typescript-operations": "^1.18.2",
     "@graphql-codegen/typescript-rtk-query": "^1.1.1",
```
Comment (JFrankfurt) [inline package.json:69]: It looks like you've removed these from the codegen file--I was surprised to see that. Do we not need `@graphql-codegen/typescript` or  `@graphql-codegen/typescript-operations`?

File: src/fetchGraphQL.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+const fetchQuery = (operation: any, variables: Variables): ObservableFromValue<GraphQLResponse> => {
+  // TODO: figure out why this returns null
+  const chainId = (store.getState() as AppState).application.chainId
+
+  const subgraphUrl = CHAIN_SUBGRAPH_URL[1]
+
+  const body = JSON.stringify({
+    query: operation.text, // GraphQL text from input
+    variables,
+  })
```
Comment (JFrankfurt) [inline src/fetchGraphQL.ts:42]: looking at the first example link there's a query name provided here. Why is that not relevant here?

File: src/fetchGraphQL.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+const headers = {
+  Accept: 'application/json',
+  'Content-type': 'application/json',
+}
+
+// Define a function that fetches the results of a request (query/mutation/etc)
+// and returns its results as a Promise:
+const fetchQuery = (operation: any, variables: Variables): ObservableFromValue<GraphQLResponse> => {
```
Comment (JFrankfurt) [inline src/fetchGraphQL.ts:33]: How should `operation` be typed here?

File: src/hooks/usePoolTickData.ts [hunk 4]
```diff
@@ -162,9 +155,7 @@ function useTicksFromSubgraph(
         )
       : undefined
 
-  return useAllV3TicksQuery(poolAddress ? { poolAddress: poolAddress?.toLowerCase(), skip: 0 } : skipToken, {
-    pollingInterval: ms`30s`,
-  })
+  return useAllV3TicksQuery(poolAddress, 0)
```
Comment (JFrankfurt) [inline src/hooks/usePoolTickData.ts:158]: looks like we're loosing a polling feature here. Is there a straightforward way to preserve this?

File: codegen.yml [hunk 5]
```diff
@@ -2,10 +2,6 @@ overrideExisting: true
 schema: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v3'
 documents: 'src/**/!(*.d).{ts,tsx}'
 generates:
-  ./src/state/data/generated.ts:
+  ./schema/schema.graphql:
```
Comment (JFrankfurt) [inline codegen.yml:5]: The old path was intentionally omitted from version control w/ .gitignore

It's generally best-practice to omit generated code from git. You should update the path to ignore the generated files here as well.

File: package.json (hunk 1 above)
Comment (cartcrom) [reply package.json:69]: Relay generates the types, so now we only need codegen to grab the schema

File: src/fetchGraphQL.ts (hunk 2 above)
Comment (cartcrom) [reply src/fetchGraphQL.ts:42]: Comment next to the name variable in that example says "//used by graphql mock on tests", I don't think the backend has any use of a name variable, that repo probably just has mocks implemented that return based on name

File: src/hooks/graphql/AllV3TicksQuery.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+`
+
+type TickData = { error: any | null; isLoading: boolean; data: AllV3TicksQuery$data | null }
+
+export default function useAllV3TicksQuery(poolAddress: string | undefined, skip: number, interval: number) {
+  const [data, setData] = useState<TickData>({ error: null, isLoading: true, data: null })
+  const chainId = useAppSelector((state) => state.application.chainId)
+  const environment = useRelayEnvironment()
+
+  const refreshData = useCallback(() => {
```
Comment (vm) [inline src/hooks/graphql/AllV3TicksQuery.ts:30]: don't think you need to `useCallback` something that's passed into `useEffect`, since it's gonna get re-run anyways. that's for making sure the value doesn't change as a prop.

File: src/fetchGraphQL.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+const headers = {
+  Accept: 'application/json',
+  'Content-type': 'application/json',
+}
+
+// Define a function that fetches the results of a request (query/mutation/etc)
+// and returns its results as a Promise:
+const fetchQuery = (params: RequestParameters, variables: Variables): ObservableFromValue<GraphQLResponse> => {
+  // TODO: figure out why this returns null
```
Comment (JFrankfurt) [inline src/fetchGraphQL.ts:34]: is this todo still outstanding?

File: src/fetchGraphQL.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  const chainId = (store.getState() as AppState).application.chainId
+
+  const subgraphUrl = CHAIN_SUBGRAPH_URL[chainId ?? SupportedChainId.MAINNET]
```
Comment (JFrankfurt) [inline src/fetchGraphQL.ts:37]: what happens when chainid is a value that The Graph doesn't support? (i.e., optimistic kovan?)

File: src/hooks/graphql/AllV3TicksQuery.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+      price1
+    }
+  }
+`
+
+type TickData = { error: any | null; isLoading: boolean; data: AllV3TicksQuery$data | null }
+
+export default function useAllV3TicksQuery(poolAddress: string | undefined, skip: number, interval: number) {
+  const [data, setData] = useState<TickData>({ error: null, isLoading: true, data: null })
+  const chainId = useAppSelector((state) => state.application.chainId)
```
Comment (JFrankfurt) [inline src/hooks/graphql/AllV3TicksQuery.ts:27]: double checking--is this the right chainId to use?

File: src/hooks/graphql/AllV3TicksQuery.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      })
+    }
+  }
+
+  useEffect(refreshData, [refreshData, poolAddress, skip])
+
+  useInterval(
+    () => {
+      refreshData()
+    },
```
Comment (JFrankfurt) [inline src/hooks/graphql/AllV3TicksQuery.ts:47]: p sure you don't need to wrap this call in the anon fn. Can just be `useInterval(refreshData, interval, true)`

File: src/hooks/graphql/FeeTierDistributionQuery.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+type FeeTierDistribitionData = { error: any | null; isLoading: boolean; data: FeeTierDistributionQuery$data | null }
+
+export default function useFeeTierDistributionQuery(
+  token0: string | undefined,
+  token1: string | undefined,
+  interval: number
+) {
+  const [data, setData] = useState<FeeTierDistribitionData>({ error: null, isLoading: true, data: null })
+  const environment = useRelayEnvironment()
+  const chainId = useAppSelector((state) => state.application.chainId)
```
Comment (JFrankfurt) [inline src/hooks/graphql/FeeTierDistributionQuery.ts:49]: same q re: appropriate chainId

File: src/hooks/graphql/FeeTierDistributionQuery.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+  useEffect(refreshData, [refreshData, token0, token1])
+
+  useInterval(
+    () => {
+      refreshData()
+    },
+    interval,
+    true
+  )
```
Comment (JFrankfurt) [inline src/hooks/graphql/FeeTierDistributionQuery.ts:71]: same comment on the anon fn wrapper

File: src/utils/computeSurroundingTicks.ts [hunk 13]
```diff
@@ -10,7 +10,7 @@ export default function computeSurroundingTicks(
   token0: Token,
   token1: Token,
   activeTickProcessed: TickProcessed,
-  sortedTickData: TickData[],
+  sortedTickData: readonly TickData[],
```
Comment (JFrankfurt) [inline src/utils/computeSurroundingTicks.ts:13]: what's the purpose of this change?

**Review by JFrankfurt — CHANGES_REQUESTED** [review]

File: src/fetchGraphQL.ts (hunk 8 above)
Comment (cartcrom) [reply src/fetchGraphQL.ts:37]: When on an unsupported chain, the pools page will not render inputs that lead to the API being called. This functions the same way the redux data store's calls functioned before, but I'll add an additional check to the function

File: src/hooks/graphql/AllV3TicksQuery.ts (hunk 10 above)
Comment (cartcrom) [reply src/hooks/graphql/AllV3TicksQuery.ts:47]: Oh duh, thanks

File: src/utils/computeSurroundingTicks.ts (hunk 13 above)
Comment (cartcrom) [reply src/utils/computeSurroundingTicks.ts]: relay's generated data type AllV3TicksQuery$data returns a readonly. I deleted this type and refactored to use the generated types instead

**Review by JFrankfurt — APPROVED** [review]

File: src/RelayEnvironment.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import { Environment, Network, RecordSource, Store } from 'relay-runtime'
```
Comment (vm) [inline src/RelayEnvironment.ts:1]: wonder if all this should live under `src/relay` rather than spread out

Comment (vm) [reply src/RelayEnvironment.ts:1]: Screen Shot 2022-08-16 at 2 52 55 PM
